torchtune/utils/_profiler.py

```diff
@@ -4,47 +4,350 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 
-import contextlib
 
-from typing import ContextManager, Optional
+import os
+import time
+from functools import partial
+from pathlib import Path
+from typing import Optional, Tuple
 
 import torch
-from torch.profiler import profile
+import torch.distributed
 
+from omegaconf import DictConfig
+from torch._C._profiler import _ExperimentalConfig
+from torch.profiler import tensorboard_trace_handler
+from torchtune.utils import get_world_size_and_rank
 
-def profiler(
-    enabled: Optional[bool] = False,
-    output_dir: Optional[str] = "./torchtune_perf_tracing.json",
-) -> ContextManager:
+from torchtune.utils.logging import get_logger
+
+log = get_logger("INFO")
+
+PROFILER_KEY = "profiler"
+_DEFAULT_PROFILER_ACTIVITIES = {
+    torch.profiler.ProfilerActivity.CPU,
+    torch.profiler.ProfilerActivity.CUDA,
+}
+
+_DEFAULT_SCHEDULE: dict = {
+    "wait": 5,
+    "warmup": 5,
+    "active": 2,
+    "repeat": 1,
+}
+
+DEFAULT_TRACE_OPTS: dict = {
+    "profile_memory": False,
+    "with_stack": False,
+    "record_shapes": True,
+    "with_flops": False,
+}
```

Comment:
I notice you use `DEFAULT_TRACE_OPTS` in the recipe to set defaults but not `_DEFAULT_SCHEDULE`. Is this just because it's more natural to use bools for default values than ints? My question is kinda similar to my other comment below.. ideally we can handle this consistently across the default values (I don't want to have half defined in the recipe and half defined in here).